Refactor 01-backup

douglatornell opened the pull request:
Split the monolithic 01-backup lesson section into 8 smaller topic sections.

I did a lot of this refactoring for a workshop that I created in Jun-2015 based on the SWC material. it is rendered at http://douglatornell.ca/training/meopeers-2015-06-15/

We are going to teach this refactored version of the hg lesson next week at the UBC EOAS workshop, so it will get trial by fire, but I would also appreciate reviewer comments.

douglatornell committed:
Rebuild HTML re: changes pulled from upstream swc repo.

douglatornell committed:
Improve wording of intro text based on Git lesson intro.

douglatornell committed:
Move version control concepts intro to 01-basics section.

Use wording and figures from git-novice lesson.

douglatornell committed:
Move Mercurial configuration into 02-configuration section.

douglatornell committed:
Move Mercurial repo creation into 03-create-repo section.

douglatornell committed:
Fix typo.

douglatornell committed:
Delete repetition of learing goals at end of lesson.

douglatornell committed:
Move adding and committing files into 04-tracking section.

douglatornell committed:
Move changing files & examining diffs to 05-changes section.

douglatornell committed:
Move bio repo challenge from 04-tracking to 05-changes section.

douglatornell committed:
Move exploring history with `hg diff` to 06-history section.

douglatornell committed:
Move recoevering old version & hg command aliases into 07-revert section.

douglatornell committed:
Move explanation of .hgignore file into 08-ignore section.

douglatornell committed:
Delete 01-backup section; replaced by 01-basics through 08-ignore sections.

jdcorless commented on instructors.html:
```diff
@@ -28,15 +28,22 @@
+<ul>
+<li>Version control’s advantages:
 <ul>
 <li>Nothing that is committed to version control is ever lost. This means it can be used like the “undo” feature in an editor, and since all old versions of files are saved it’s always possible to go back in time to see exactly who wrote what on a particular day, or what version of a program was used to generate a particular set of results.</li>
```
I think the clause "it’s always possible to go back in time to see exactly who wrote what on a particular day" is redundant with the next list item

douglatornell replied:
I'll include this in the rewrite of 1`instructors.md`.

jdcorless commented on 02-configuration.md:
```diff
@@ -0,0 +1,99 @@
+The command
+
+~~~ {.bash}
+$ hg config --edit
```
On a new Ubuntu VM with hg ver 2.8.2 installed, this command did not work (hg: unknown command 'config').  I think in the old hg lesson we just had students editing the config file manually.

douglatornell replied:
Good catch! Looks like `hg config --edit` was introduced in Mercurial v3.0 (2014-05-01).

Yes, the previous version had direct editing of the config file. `hg config --edit` avoids the diversion of where the config file is on different platforms and what it is called. Would it work to put the direct editing approach in a callout to be used in the cases where `hg config --edit` fails?

jdcorless replied:
I think that makes sense, especially if your install instructions get most users set up with hg v3.0 and then the callout is just for those who decided not to upgrade.

jdcorless commented on 02-configuration.md:
```diff
@@ -0,0 +1,99 @@
+
+Edit the file to set `username` to your own name and email address and add your
+favourite editor in the `[ui]` section.
+Also uncomment the the `pager =`,
```
typo

jdcorless commented on 02-configuration.md:
```diff
@@ -0,0 +1,99 @@
+You can check that your settings are the way that you want them with the command
+
+~~~ {.bash}
+$ hg config ui extensions
```
Same issue (hg: unknown command 'config')

jdcorless commented on 02-configuration.md:
```diff
@@ -0,0 +1,99 @@
+directory means that they will be used for every Mercurial repository on this
+machine.
+
+The above configuration work only needs to be done tghe first time you use
```
typo

jdcorless commented on 04-tracking.md:
```diff
@@ -0,0 +1,145 @@
+(or whatever other editor we configured at the start)
+so that we can write a longer message.
+
+[Good commit messages][commit-messages] start with a brief (<70 characters) summary of
```
Cool idea to link to this as reference material for students!

douglatornell replied:
Credit to the git-novice lesson for reminding me of that link in this round of edits, though I know I've liked to it in other material.

On the topic of reference links, I like to include a richer collection on the index page than the generic SWC lesson does (Reference, Discussion, Instructor's Guide). I usually add the Mercurial site, the Mercurial - The Definitive Guide book site, and a collection of "[command reminders](https://douglatornell.github.io/hg-novice-2015-09-22-ubc/commands.html)". Your comment makes me think that the Good Commit Messages link would be a good addition to that.

What do you think about making that collection of links a standard thing on the index page for the Mercurial lesson?

jdcorless replied:
I really like that idea.  I think that we want to leave students with essentially a curated list of where to go to build on the intro information they've just gone through.  Web searching is great but nothing like an expert encouraging you in the right direction!

kynan commented on 01-basics.md:
```diff
@@ -0,0 +1,50 @@
+
+"Piled Higher and Deeper" by Jorge Cham, http://www.phdcomics.com
+
+We've all been in this situation before: it seems ridiculous to have multiple nearly-identical versions of the same document. Some word processors let us deal with this a little better, such as Microsoft Word's "Track Changes" or Google Docs' version history.
```
Minor niggle: would be good to be consistent about line length: either wrap or don't

kynan commented on 01-basics.md:
```diff
@@ -0,0 +1,50 @@
+
+![Different versions can be saved](fig/versions.svg)
+
+If there aren't conflicts, you can even try to play two sets of changes onto the same base document.
```
are no conflicts

kynan commented on 04-tracking.md:
```diff
@@ -0,0 +1,145 @@
+but please don't use a word processor like Microsoft Word or LibreOffice Write.
+In particular,
+the editor does not have to be the one that you set in your Mercurial
+configuration earlier.)
```
Extra )

kynan commented on 07-revert.md:
```diff
@@ -0,0 +1,118 @@
+> (no backup revert) that is the same as typing `hg revert --no-backup`.
+> Note that it is good practice when creating aliases for commands to give them
+> your own names instead of using the alias to redefine the built-in command.
+If we had used the alias to change add the `--no-backup` flag to the `revert`
```
`>` missing

kynan commented on index.md:
```diff
@@ -17,17 +17,30 @@ can and should be stored in a version control system.
+*   As we have this record of who made what changes when,
+    we know who to ask if we have questions later on,
+    and,
+    if needed it,
```
remove "it"

kynan commented:
Apart from minor niggles this is looking good I think.

2 general remarks:
- It would be good to be consistent about line length for paragraphs: either break at a fix margin (e.g. 80 columns) or not at all.
- Do we want to use the informal style (let's, don't) as it is now?

douglatornell committed:
Add links to Mercurial site.

douglatornell committed:
Break overly long lines.

As recommended by @kynan in PR #22 review.

douglatornell committed:
Fix typos, etc. spotted by @kynan in review of PR #22.

douglatornell committed:
Fix typos spotted by @jdcorless in review of PR #22.

douglatornell committed:
Put the 'committing only some files' text into a callout.

Unless it's in a callout, it looks confusingly as if the student
should be typing these commands.

Contributed by @nxg in PR #24.

douglatornell committed:
Add callout containing config instructions if `hg config` fails.

Versions of Mercurial <3.0 (May-2014) don't have the `hg config`
so users have to create the appropriate file manually.
See discussion with @jdcorless in PR #22.

kynan commented:
@douglatornell: seems this has some conflicts now.

douglatornell commented:
Yeah, I noticed that when I was working on it last week. Will try to
resolve them this week.

Does anyone have any advice/opinions on whether I should merge this into
gh-pages, or into a new working branch for the v5.4 lesson release? I have
more material coming for the v5.4 release, and would like to have the
complete new lesson available somewhere before the 13-Jun Bug BBQ. But I'm
not sure if the material should go into the gh-pages branch before then.

On Tue, May 17, 2016 at 1:30 PM Florian Rathgeber notifications@github.com
wrote:

> @douglatornell https://github.com/douglatornell: seems this has some
> conflicts now.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/swcarpentry/hg-novice/pull/22#issuecomment-219843926

gvwilson commented:
I suggest putting it into gh-pages.

douglatornell commented:
Okay. No worries about messing up instructor/workshops between now and the
v5.4 release?

On Tue, May 17, 2016 at 2:34 PM Greg Wilson notifications@github.com
wrote:

> I suggest putting it into gh-pages.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/swcarpentry/hg-novice/pull/22#issuecomment-219861456

gvwilson commented:
I don't think we have any Mercurial workshops scheduled between now and 
then.

kynan commented on 02-configuration.md:
```diff
@@ -0,0 +1,131 @@
+You can check that your settings are the way that you want them with the command
+
+~~~ {.bash}
+$ hg config ui extensions
```
This will again fail in hg pre 3.0

kynan commented on 01-basics.md:
```diff
@@ -0,0 +1,65 @@
+> [Git](http://swcarpentry.github.io/git-novice/).
+> In particular,
+> the latter are *distributed*, meaning that they don't need a centralized
+> server to host the repository.
```
Instead, every user has their own full copy of the repository.
